pkg/hooks/hooks.go

```diff
@@ -1,141 +1,104 @@
+	manager = &Manager{
+		hooks:       map[string]*current.Hook{},
+		directories: directories,
 	}
-	if _, err := os.Stat(hook.Hook); err != nil {
-		return hook, errors.Wrapf(err, "unable to stat hook %q in hook config %q", hook.Hook, hookPath)
-	}
-	validStage := map[string]bool{"prestart": true, "poststart": true, "poststop": true}
-	for _, cmd := range hook.Cmds {
-		if _, err = regexp.Compile(cmd); err != nil {
-			return hook, errors.Wrapf(err, "invalid cmd regular expression %q defined in hook config %q", cmd, hookPath)
-		}
-	}
-	for _, cmd := range hook.Annotations {
-		if _, err = regexp.Compile(cmd); err != nil {
-			return hook, errors.Wrapf(err, "invalid cmd regular expression %q defined in hook config %q", cmd, hookPath)
-		}
-	}
-	for _, stage := range hook.Stage {
-		if !validStage[stage] {
-			return hook, errors.Wrapf(err, "unknown stage %q defined in hook config %q", stage, hookPath)
-		}
-	}
-	return hook, nil
-}
 
-// readHooks reads hooks json files in directory to setup OCI Hooks
-// adding hooks to the passedin hooks map.
-func readHooks(hooksPath string, hooks map[string]HookParams) error {
-	if _, err := os.Stat(hooksPath); err != nil {
-		if os.IsNotExist(err) {
-			logrus.Warnf("hooks path: %q does not exist", hooksPath)
-			return nil
+	for _, dir := range directories {
+		err = ReadDir(dir, manager.hooks)
+		if err != nil {
+			return nil, err
 		}
-		return errors.Wrapf(err, "unable to stat hooks path %q", hooksPath)
 	}
 
-	files, err := ioutil.ReadDir(hooksPath)
-	if err != nil {
-		return err
-	}
+	return manager, nil
+}
 
-	for _, file := range files {
-		if !strings.HasSuffix(file.Name(), ".json") {
-			continue
-		}
-		hook, err := readHook(filepath.Join(hooksPath, file.Name()))
+// Hooks injects OCI runtime hooks for a given container configuration.
+func (m *Manager) Hooks(config *rspec.Spec, annotations map[string]string, hasBindMounts bool) (err error) {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+	for name, hook := range m.hooks {
+		match, err := hook.When.Match(config, annotations, hasBindMounts)
 		if err != nil {
-			return err
+			return errors.Wrapf(err, "matching hook %q", name)
 		}
-		for key, h := range hooks {
-			// hook.Hook can only be defined in one hook file, unless it has the
-			// same name in the override path.
-			if hook.Hook == h.Hook && key != file.Name() {
-				return errors.Wrapf(syscall.EINVAL, "duplicate path,  hook %q from %q already defined in %q", hook.Hook, hooksPath, key)
+		if match {
+			if config.Hooks == nil {
+				config.Hooks = &rspec.Hooks{}
+			}
+			for _, stage := range hook.Stages {
+				switch stage {
+				case "prestart":
+					config.Hooks.Prestart = append(config.Hooks.Prestart, hook.Hook)
+				case "poststart":
+					config.Hooks.Poststart = append(config.Hooks.Poststart, hook.Hook)
+				case "poststop":
+					config.Hooks.Poststop = append(config.Hooks.Poststop, hook.Hook)
+				default:
+					panic(fmt.Sprintf("unknown stage %q", stage))
```

Comment:
Shouldn't this be a return err.  Panic would bring cri-o down.

Reply:
Agree, should probably have an ErrUnknownStage and `return errors.Wrapf(ErrUnknownStage, "invalid stage %q", stage)` or similar

Reply:
> Shouldn't this be a return err. Panic would bring cri-o down.

This should be impossible, because we [validate stages][1] [after reading them off the disk][2].  The `Manager` hook information [is private]][3], so I'm not aware of a way to get around the post-read validation and inject a buggy hook.  So I don't think the `panic` is an issue, but I'll push an update anyway.  There could always be bugs I've missed ;).

[1]: https://github.com/projectatomic/libpod/pull/686/files#diff-697f3ecb2c90bb150662f21f7d0ee75fR78
[2]: https://github.com/projectatomic/libpod/pull/686/files#diff-ffa06af0c211f9169abc4254595ebc79R39
[3]: https://github.com/projectatomic/libpod/pull/686/files#diff-4ce0c4408808dad0b46844b92388f205R28

Reply:
> Agree, should probably have an ErrUnknownStage and `return errors.Wrapf(ErrUnknownStage, "invalid stage %q", stage)` or similar

Somehow I missed this comment earlier.  I've updated this line since to return an error, but have not defined an `ErrUnknownStage`.  Do you expect callers to want to switch on that vs. other error conditions?  I'm having trouble imagining when you'd want different handling than for, say, [invalid-regexp errors][1].

[1]: https://github.com/projectatomic/libpod/pull/686/files#diff-a483e4864f7af4b01cb5c7656ab79ddbR24

Reply:
If we don't see this being hit often at all (and given your earlier comment on validation elsewhere, I'm content to believe that), there's no real point to having a specific error -- if we didn't have such validation, I think it would be helpful to have an identifier for a potentially common error